src/linear_algebra/charpoly.lean

```diff
@@ -0,0 +1,138 @@
+    by rw [basis_to_matrix_mul_linear_map_to_matrix_mul_basis_to_matrix]
+  ... = (scalar ι' X - C.map_matrix (φ₁ P ⬝ φ₂ A' ⬝ φ₃ Q)).det :
+    by rw [reindex_linear_equiv_mul R R _ _ e, reindex_linear_equiv_mul R R e _ _]
+  ... = (scalar ι' X - (C.map_matrix (φ₁ P) ⬝ C.map_matrix A' ⬝ C.map_matrix (φ₃ Q))).det : by simp
+  ... = (scalar ι' X ⬝ C.map_matrix (φ₁ P) ⬝ (C.map_matrix (φ₃ Q)) -
+    (C.map_matrix (φ₁ P) ⬝ C.map_matrix A' ⬝ C.map_matrix (φ₃ Q))).det :
+      by { rw [matrix.mul_assoc ((scalar ι') X), hPQ, matrix.mul_one] }
+  ... = (C.map_matrix (φ₁ P) ⬝ scalar ι' X ⬝ (C.map_matrix (φ₃ Q)) -
+    (C.map_matrix (φ₁ P) ⬝ C.map_matrix A' ⬝ C.map_matrix (φ₃ Q))).det : by simp
+  ... = (C.map_matrix (φ₁ P) ⬝ (scalar ι' X - C.map_matrix A') ⬝ C.map_matrix (φ₃ Q)).det :
+    by rw [← matrix.sub_mul, ← matrix.mul_sub]
+  ... = (C.map_matrix (φ₁ P)).det * (scalar ι' X - C.map_matrix A').det *
+    (C.map_matrix (φ₃ Q)).det : by rw [det_mul, det_mul]
+  ... = (C.map_matrix (φ₁ P)).det * (C.map_matrix (φ₃ Q)).det *
+    (scalar ι' X - C.map_matrix A').det : by ring
+  ... = (scalar ι' X - C.map_matrix A').det : by rw [← det_mul, hPQ, det_one, one_mul]
+  ... = f.charpoly : rfl
+end
+
+end basic
+
+section coeff
+
+lemma charpoly_monic : f.charpoly.monic := charpoly_monic _
+
+end coeff
+
+section cayley_hamilton
+
+/-- The Cayley-Hamilton Theorem, that the characteristic polynomial of a linear map, applied to
+the linear map itself, is zero. -/
+lemma aeval_self_charpoly : aeval f f.charpoly = 0 :=
+begin
+  apply (linear_equiv.map_eq_zero_iff (alg_equiv_matrix _).to_linear_equiv).1,
+  rw [alg_equiv.to_linear_equiv_apply, ← alg_equiv.coe_alg_hom,
+    ← polynomial.aeval_alg_hom_apply _ _ _, charpoly_def],
+  exact aeval_self_charpoly _,
+end
+
+lemma is_integral : is_integral R f := ⟨f.charpoly, ⟨charpoly_monic f, aeval_self_charpoly f⟩⟩
+
+lemma minpoly_dvd_charpoly {K : Type u} {M : Type v} [field K] [add_comm_group M] [module K M]
+  [finite_dimensional K M] (f : M →ₗ[K] M) : minpoly K f ∣ f.charpoly :=
+minpoly.dvd _ _ (aeval_self_charpoly f)
+
+variable {f}
+
+lemma charpoly_coeff_zero_of_injective (hf : function.injective f) : (minpoly R f).coeff 0 ≠ 0 :=
+begin
+  intro h,
+  obtain ⟨P, hP⟩ := X_dvd_iff.2 h,
+  have hdegP : P.degree < (minpoly R f).degree,
+  { rw [hP, mul_comm],
+    refine degree_lt_degree_mul_X (λ h, _),
+    rw [h, mul_zero] at hP,
+    exact minpoly.ne_zero (is_integral f) hP },
+  have hPmonic : P.monic,
+  { suffices : (minpoly R f).monic,
+    { rwa [monic.def, hP, mul_comm, leading_coeff_mul_X, ← monic.def] at this },
+    exact minpoly.monic (is_integral f) },
+  have hzero : aeval f (minpoly R f) = 0 := minpoly.aeval _ _,
+  simp only [hP, mul_eq_comp, ext_iff, hf, aeval_X, map_eq_zero_iff, coe_comp, alg_hom.map_mul,
+    zero_apply] at hzero,
+  exact not_le.2 hdegP (minpoly.min _ _ hPmonic (ext hzero)),
+end
+
+end cayley_hamilton
+
+end linear_map
```

Comment:
A bit annoying that `End` doesn't get unfolded enough to make eigenspace.lean:144 work, but i guess it's fine to re-define it here:
```suggestion
end linear_map

lemma module.End.is_integral : is_integral R f := f.is_integral
```

Reply:
It does! I did a mistake resolving a `git` conflict and I removed `import linear_algebra.charpoly`. Feel free to sort the imports if you want :)